snuba/subscriptions/scheduler_processing_strategy.py

```diff
@@ -0,0 +1,150 @@
+    """
+    The TickBuffer buffers ticks until they are ready to be submitted to
+    the next processing step.
+
+    The behavior of the TickBuffer depends on which of the two scheduler
+    modes applies.
+
+    If the scheduler mode is IMMEDIATE then there is no buffering and a
+    message is always immediately submitted to the next processing step.
+
+    If the scheduler mode is WAIT_FOR_SLOWEST_PARTITION then messages are
+    buffered until all partitions are at least up to the timestamp of the
+    tick.
+
+    `max_ticks_buffered_per_partition` applies if the scheduler mode is
+    WAIT_FOR_SLOWEST_PARTITION. Once the maximum ticks is received for that
+    partition, we start to submit ticks for processing even if that timestamp
+    is not received for all partitions yet.
+
+    This prevents the buffered tick list growing infinitely and scheduling
+    to grind to a halt if one partition starts falling far behind for some reason.
+    """
+
+    def __init__(
+        self,
+        mode: SchedulerMode,
+        partitions: int,
+        max_ticks_buffered_per_partition: Optional[int],
+        next_step: ProcessingStrategy[Tick],
+        metrics: MetricsBackend,
+    ) -> None:
+        if mode == SchedulerMode.WAIT_FOR_SLOWEST_PARTITION:
+            assert max_ticks_buffered_per_partition is not None
+
+        self.__mode = mode
+        self.__partitions = partitions
+        self.__max_ticks_buffered_per_partition = max_ticks_buffered_per_partition
+        self.__next_step = next_step
+        self.__metrics = metrics
+
+        self.__buffers: Mapping[int, Deque[Message[Tick]]] = {
+            index: deque() for index in range(self.__partitions)
+        }
+
+        # Stores the latest timestamp we received for any partition. This is
+        # just for recording the partition lag.
+        self.__latest_ts: Optional[datetime] = None
+
+    def poll(self) -> None:
+        pass
+
+    def submit(self, message: Message[Tick]) -> None:
+        # If the scheduler mode is immediate or there is only one partition
+        # immediately submit message to the next step.
+        # We don't keep any latest_ts values as it is not relevant.
+        if self.__mode == SchedulerMode.IMMEDIATE or self.__partitions == 1:
+            self.__next_step.submit(message)
+            return
+
+        # Update the latest_ts for metrics
+        if (
+            self.__latest_ts is None
+            or message.payload.timestamps.upper > self.__latest_ts
+        ):
+            self.__latest_ts = message.payload.timestamps.upper
+
+        tick_partition = message.payload.partition
+        self.__buffers[tick_partition].append(message)
+
+        # If the buffer length exceeds `max_ticks_buffered_per_partition`
+        # immediately submit the earliest message in that buffer to the next step.
+        if len(self.__buffers[tick_partition]) > cast(
+            int, self.__max_ticks_buffered_per_partition
+        ):
```

Comment:
Is there any advantage tracking the maximum amount of ticks per partition in terms of number of commit messages instead of time? IT seems that if we set a limit in terms of time, the behavior here would be independent from the granularity of the commit messages. With this current strategy, instead, the bigger the batch we set the more sparse the commit messages, thus the longer it would take to identify a late partition.

Reply:
Ticks seemed the most straightforward to me since it's number of ticks (not time between them) that are the ones actually filling are buffers and potentially causing problems if we have too many. I think having a max time as well might also make sense but to me that more of a product concern about when we want to call a partition "too late". My primary motivation in limiting the number of ticks is only to prevent infinite growth of buffers, and not really about the product behaviour which I think having the max in terms of time addresses.

Reply:
So if the concern is the buffer size (fair), the maximum should be a setting and not an argument we pass to the consumer as it is not related to the product functionality, or to ensure we make progress. It is meant to just preserve the integrity of the consumer itself.